src/spawn/spawn_server.rs

```diff
@@ -0,0 +1,427 @@
+            }
+
+            file_actions_op_p = unsafe {
+                let faop = *file_actions_op_p;
+                faop.next
+            };
+        }
+    }
+
+    // let c_str = CString::new("TARGET PID: %d\n").unwrap();
+    // let c_str_2 = CString::new("TARGET PPID: %d\n").unwrap();
+    // unsafe { libc::printf(c_str.as_ptr(), libc::getpid()); };
+    // unsafe { libc::printf(c_str_2.as_ptr(), libc::getppid()); };
+
+    // unsafe { libc::execvpe(executable, ptr_array, envp.as_ptr()) };
+    let spawn_func = if use_path {
+        libc::posix_spawn
+    } else {
+        libc::posix_spawnp
+    };
+
+    
+    if unsafe { spawn_func(
+        pid, executable, file_actions_p, spawn_attrs, argv, envp
+    ) } < 0 {
+        panic!("posix_spawn: {}", std::io::Error::last_os_error())
+    };
+
+
+    if unsafe { *pid } < 0 {
+        panic!("posix_spawn: {}", std::io::Error::last_os_error())
+    }
+
+    // communicate created PID back to calling process, which communicates again back to client
+    nix::unistd::close(target_pid_pipe[0]).unwrap();
+    unsafe {
+        if libc::write(
+            target_pid_pipe[1],
+            pid as *mut c_void,
+            std::mem::size_of::<*mut pid_t>()
+        ) as usize != std::mem::size_of::<*mut pid_t>() {
+            panic!("write: '{}'", std::io::Error::last_os_error());
+        }
+    };
+
+    // let cstr1 = CString::new("TARGET PID %d\n").unwrap();
+    // let cstr2 = CString::new("TARGET PARENT PID %d\n").unwrap();
+
+    // let pid1 = unsafe { libc::getpid() };
+    // let pid2 = unsafe { libc::getppid() };
+
+    // unsafe { libc::printf(cstr1.as_ptr(), pid1); };
+    // unsafe { libc::printf(cstr2.as_ptr(), pid2); };
+
+    0
+}
+
+
+fn exec_posix_spawn(spawn_args: ApiPosixSpawnArgs) -> (libc::pid_t, libc::pid_t) {
+    let ApiPosixSpawnArgs {
+        executable,
+        file_actions, // Not implementing for now; de-/serializing is not trivial
+        spawn_attrs: spawnattr_t,  // Not implementing for now; de-/serializing is not trivial
+        argv,
+        envp,
+        use_path,
+        client_pid,
+    } = spawn_args;
+
+    let mut stack: [libc::c_int; 1024] = [0; 1024];
+    let spawner_pid: pid_t = 0;
+    let mut target_pid: pid_t = 0;
+    let mut pipe_fds: [libc::c_int; 2] = [0i32; 2];
+
+    let executable = CString::new(executable).unwrap();
+    let file_actions =  unsafe { std::mem::transmute::<[u8; 80], libc::posix_spawn_file_actions_t>(file_actions) };
+    let spawn_attrs = unsafe { std::mem::transmute::<[u8; 336], libc::posix_spawnattr_t>(spawnattr_t) };
+
+    // Correctly allocate and populate argv
+    let mut argv: Vec<*mut c_char> = argv.iter().map(|arg| arg.as_ptr() as *mut c_char).collect();
+    argv.push(std::ptr::null_mut());
+
+    // Correctly allocate and populate envp
+    let mut envp: Vec<*mut c_char> = envp.iter().map(|var| var.as_ptr() as *mut c_char).collect();
+    envp.push(std::ptr::null_mut());
```

Comment:
This is currently broken.

You need to pass null terminated strings, so a conversion to `CString`s is necessary before passing them on to `clone()`.

This causes faulty arguments and environment variables to be processed by the child.

Have a look at the following code I tested locally:
```rust
    // Keep an array of owned CStrings to prevent them from being dropped to early
    // Drop them *after* forking to prevent invalid data in the child
    // TODO: probaly not unwrap here (will fail if string contains null bytes)
    let argv_cstrings: Vec<CString> = argv
        .into_iter()
        .map(|arg| CString::new(arg).unwrap())
        .collect();
    let envp_cstrings: Vec<CString> = envp
        .into_iter()
        .map(|arg| CString::new(arg).unwrap())
        .collect();

    // Correctly allocate and populate argv
    let mut argv: Vec<*mut c_char> = argv_cstrings
        .iter()
        .map(|arg| arg.as_ptr() as *mut c_char)
        .collect();
    argv.push(std::ptr::null_mut());

    // Correctly allocate and populate envp
    let mut envp: Vec<*mut c_char> = envp_cstrings
        .iter()
        .map(|var| var.as_ptr() as *mut c_char)
        .collect();
    envp.push(std::ptr::null_mut());


```